wrappers/azurekeyvault/azurekeyvault.go

```diff
@@ -261,26 +276,318 @@ func (v *Wrapper) Decrypt(ctx context.Context, in *wrapping.EncryptedBlobInfo, a
+
+// keyTransferBlob represents the structure of a key transfer blob
+// used to import a key into Azure Key Vault. The format of the key
+// transfer blob uses JSON Web Encryption compact serialization [RFC7516]
+// primarily as a vehicle for delivering the required metadata to the
+// service for correct decryption.
+type keyTransferBlob struct {
+	SchemaVersion string                `json:"schema_version"`
+	Header        keyTransferBlobHeader `json:"header"`
+	CipherText    string                `json:"ciphertext"`
+	Generator     string                `json:"generator"`
+}
+
+// keyTransferBlobHeader represents the header of a key transfer blob.
+type keyTransferBlobHeader struct {
+	Kid string `json:"kid"`
+	Alg string `json:"alg"`
+	Enc string `json:"enc"`
+}
+
+// jwkToRSAPublicKey returns a pointer to an rsa.PublicKey that contains
+// the modulus and public exponent from the given keyvault.JSONWebKey (JWK).
+// Both the RSA modulus "N" and public exponent "E" in the JWK are
+// parsed as the base64url encoding of the value's unsigned big-endian
+// representation as an octet sequence. For more details, see
+// https://tools.ietf.org/html/rfc7518#section-6.3.1
+func jwkToRSAPublicKey(jwk *keyvault.JSONWebKey) (*rsa.PublicKey, error) {
+	if jwk == nil || jwk.N == nil || jwk.E == nil {
+		return nil, fmt.Errorf("must provide an RSA type JSONWebKey")
+	}
+
+	// Decode the RSA modulus
+	base64N := to.String(jwk.N)
+	nd, err := base64.RawURLEncoding.DecodeString(base64N)
+	if err != nil {
+		return nil, err
+	}
+	n := new(big.Int).SetBytes(nd)
+
+	// Decode the RSA public exponent
+	base64E := to.String(jwk.E)
+	ed, err := base64.RawURLEncoding.DecodeString(base64E)
+	if err != nil {
+		return nil, err
+	}
+	e := new(big.Int).SetBytes(ed)
+
+	return &rsa.PublicKey{
+		N: n,
+		E: int(e.Uint64()),
+	}, nil
+}
+
+// wrapTargetKey returns wrapped key material according to the steps outlined in
+// the Azure Key Vault BYOK specification. For details, see the specification at
+// https://docs.microsoft.com/en-us/azure/key-vault/keys/byok-specification#key-transfer-blob.
+//
+// More specific details related to the wrapping and unwrapping mechanism used can be found at
+// http://docs.oasis-open.org/pkcs11/pkcs11-curr/v2.40/cos01/pkcs11-curr-v2.40-cos01.html#_Toc408226908.
+func (v *Wrapper) wrapTargetKey(pub *rsa.PublicKey, targetKey []byte) (string, error) {
+	// Generate an ephemeral AES key
+	aesKey, err := uuid.GenerateRandomBytes(32)
+	if err != nil {
+		return "", err
+	}
+
+	// Wrap the AES key with the RSA public key using RSA-OAEP with SHA1
+	aesKeyWrapped, err := rsa.EncryptOAEP(sha1.New(), rand.Reader, pub, aesKey, []byte{})
+	if err != nil {
+		return "", err
+	}
+
+	// Wrap the target key plaintext using AES Key Wrap with Padding [RFC5649]
+	kwp, err := subtle.NewKWP(aesKey)
+	if err != nil {
+		return "", err
+	}
+	wrappedTargetKey, err := kwp.Wrap(targetKey)
+	if err != nil {
+		return "", err
+	}
+
+	// Zero the ephemeral AES key
```

Comment:
This reminds me of Vault's `memzero` func. It's a best-effort cleanup since the GC may have copied this value before it got zero'ed.

Reply:
I just read about Vault's `memzero` func. That's interesting, and I wasn't aware of that. Any recommendation related to that for this PR? It looks like best-effort is all that can be done, right?

Reply:
Yeah, not much we can do, but maybe something worth adding to the comments for clarity.

Reply:
Good idea. Added a comment in 52ad942.